### PR TITLE
Feat: Add coverage for C# applications

### LINKED ISSUE #2052: [feature]: coverage for CSharp applications
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Summary

* Just like node, python, Java and Go there should also be coverage for Csharp applications.

### Why should this be worked on?

* Csharp dependencies will be  run with coverage 

### Repository

keploy

### PR DESCRIPTION
**Pull request by yaten2302**
## What does this PR do?

This PR adds coverage for C# applications, just like python, JavaScript, Golang etc.

## Related PRs and Issues

Fixes #2052 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:
- [x] Have you read the [Contributing Guidelines on issues](https://keploy.io/docs/keploy-explained/contribution-guide/)?
- [x] My code follows the style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by yaten2302** [comment]
@shivamsouravjha , could you please have a look at the `utils.go`. All the functions which have been added are enough ig for generating and merging the coverage reports?

**Comment by shivamsouravjha** [comment]
the builds are failing please fix it.

**Comment by yaten2302** [comment]
@shivamsouravjha , I was checking the logs, why the builds are failing, almost all are failing due to missing return in [`GetCoverage()` func](https://github.com/keploy/keploy/pull/2362/files#diff-1ca089260bc8e760289fdc9354cf7dff2512760817ffb18db41f70027d4753b5R74).
But, I'm not able to understand that why're these 2 builds failing:
1. https://github.com/keploy/keploy/actions/runs/11607480978/job/32321046415?pr=2362
2. https://github.com/keploy/keploy/actions/runs/11607480990/job/32321046119?pr=2362#logs

Could you please guide regarding this?
Also, could you please guide regarding [this](https://keploy.slack.com/archives/C035URHSGDQ/p1730217077153109?thread_ts=1729510229.235849&cid=C035URHSGDQ) approach, do I've to do like this or am I going wrong somewhere?

**Comment by shivamsouravjha** [comment]
Basically in java.go we tried reading from the default output if thats the case with C# surely use the cobertura format.

**Comment by shivamsouravjha** [comment]
Ignore the second one but first one seems to be around your code changes

**Comment by yaten2302** [comment]
> Ignore the second one but first one seems to be around your code changes

@shivamsouravjha , yes the first error is due to my code changes, that error is because of the `GetCoverage()` func (as mentioned above). I'll complete that func and the error will be resolved. Currently, I'm trying that how can we read `.cobertura` files. (as in the case of Java) and similar is the case with C#

**Comment by yaten2302** [comment]
@shivamsouravjha , I was checking a sample coverage file [here](https://gist.github.com/apetro/fcfffb8c4cdab2c1061d). I've a doubt that, while returning the coverage, do I've to return the overall project line coverage or overall project branch coverage?

**Comment by shivamsouravjha** [comment]
You can refer how we handle in other functions we would want similar behaviour

**Comment by yaten2302** [comment]
@shivamsouravjha , in [these](https://github.com/keploy/keploy/blob/main/pkg/platform/coverage/java/java.go#L77-L79) lines (in `java.go`), the `totalCov` is for the complete coverage (line by line) and the `FileCov` is for calculating coverage by class, as mentioned [here](https://github.com/keploy/keploy/blob/main/pkg/platform/coverage/java/java.go#L127), am I right?
Actually, I was asking this, because according to this only, I'll create a similar func for C# applications, providing similar coverage results.

In short:
I've to give 2 result coverages i.e. `totalCov` that contains the complete line by line coverage and `FileCov` that contains class
coverage?

**Comment by shivamsouravjha** [comment]
Yes you're  right

**Comment by yaten2302** [comment]
@shivamsouravjha , I've marked this PR as ready for review, could you please have a look at it and LMK if any changes are required?

**Comment by yaten2302** [comment]
Hey @shivamsouravjha , are there any more change(s) required on this PR? Can this be merged if this looks good :)

File: pkg/platform/coverage/csharp/utils.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to merge coverage files: %w", err)
+	}
+
+	return nil
+}
+
+func GenerateCoverageReport(ctx context.Context) error {
+	reportDir := "target/site/keployE2E"
```
Comment (Sarthak160) [inline pkg/platform/coverage/csharp/utils.go:110]: Is it confirm that coverage report will be present here only ? Also is this configurable by the user ??

**Review by Sarthak160 — APPROVED** [review]
Looks Good!!
